_posts/2018-11-25-rafting-for-fun-and-profit.md

```diff
@@ -0,0 +1,442 @@
+        }
+    }
+```
+
+I marked transition changes with (!) comment.
+
+So far so good, now let's dig into something even more complex:
+
+
+```javascript
+it('should transition to candidate role when it did not receive a heartbeat', async function() {
+});
+```
+
+Hmm - this one is more tricky, we need some timeouts
+and keep track of the last heartbeat.
+I added a `setTimeout` coupled with `clearTimeout`,
+a function that will check if leader timed out - `leaderTimedOut()`
+and voila: it works...
+
+```javascript
+
+class RaftStateMachine {
+    constructor({ nodeId }) {
+        if (nodeId == undefined) {
+            throw new Error('nodeId and peers must be defined, peers must be an array');
+        }
+
+        this.state = { ...emptyState, peers: new Set(), nodeId };
+        this.alreadyVotedInThisTerm = false;
+        this.resetLeaderTimeout();
+    }
+    // ...
+
+    async transitionToCandidate() {
+        if (this.leaderTimedOut() && this.state.role !== roles.leader) {
+            this.state.role = roles.candidate;
+            this.state.currentTerm += 1;
+            this.state.alreadyVotedInThisTerm = true;
+
+            const votes = await this.getVotes();
+            const positive = this.countPositiveVotes(votes);
+            logger.log('received positive votes', positive);
+            this.state.votesGranted = positive;
+
+            if (positive > (this.state.peers.size / 2)) {
+                clearTimeout(this.leaderTimeout);
+                this.transitionToLeader();
+            } else {
+                this.resetLeaderTimeout(); // (1)
+            }
+        }
+    }
+
+    resetLeaderTimeout() {
+        clearTimeout(this.leaderTimeout);
+        this.leaderTimeout = setTimeout(this.transitionToCandidate.bind(this), config.leaderTimeout);
+    }
+
+    // ...
+}
+```
+
+But it's kind of "janky"[*](http://jankfree.org/)...
+Sometimes the test completes in milliseconds,
+sometimes it runs for way longer...
+To understand why I had to refresh my memory about event loop
+and conditions for a node process to exit.
+
+Every asynchronous *thing* is going to be queued up in the event loop
+to be processed at the later time.
+A node process will not exit if it has anything left to do in the event loop.
+I made sure that the event loop was constantly occupied
+by clearing a timeout an starting a new one every time I called `resetLeaderTimeout()`.
+
+The situation looks like this when `transitionToCandidate()` is called:
+
+```
+Stack: [transitionToCandidate()]
+Event loop: []
+```
+
+And by the time it exits it's on the event loop (1)
+
+```
+Stack: []
+Event loop: [transitionToCandidate()]
+```
+
+Ok, so how to fix it?
+There is a function that can be called on a
+`timeout`,
+`immediate`
+and `subprocess`
+called [unref](https://nodejs.org/api/timers.html#timers_timeout_unref)
+which makes "object not require the Node.js event loop to remain active".
+
+Alright, we fixed that one, but it's not ideal.
```

Comment:
Tutaj jest trochę za duży przeskok i mam wrażenie, że cały artykuł zmienił temat. Myślałem, że to o algorytmie/refactorowaniu implementacji algorytmu, a tu nagle w dużej mierze ogarnianie testów. 

Napisałbym coś, że powstały kolejne testy i poprawiona implementacja, której kod jest dostępny tu i tu. Ale pojawił się problem z czasem uruchomienia testów.